TUTORIAL.adoc

```diff
@@ -0,0 +1,511 @@
+----
+(def any-to-foo
+  (m/rewrite
+    ?x
+    "foo"))
+
+(any-to-foo "bar")
+;=> "foo"
+(any-to-foo "baz")
+;=> "foo"
+----
+
+Meander compiles a function that takes some input, and rewrites it to a new output.
+
+
+== Structural pattern matching
+
+Unlike Clojure's pervasive sequence abstraction that applies to all collections,
+Meander matches structure explicitly.
+
+[source,clojure]
+----
+(def g
+  (m/rewrite
+    [?a ?b ?c ?d]
+    {?a ?c ?b ?d}))
+
+(g [1 2 3 4])
+;=> {1 3, 2 4}
+(g '(1 2 3 4))
+;=> *fail*
+----
+
+If we look for a vector, we will only match with a vector.
+When there is a match, the output is explicitly of the type that the pattern specifies.
+
+== Unification
+
+[source,clojure]
+----
+(def a
+  (m/rewrite
+    [?x ?x]
+    true))
+
+(a [1 1])
+;=> true
+(a [1 2])
+;=> *fail*
+----
+
+Unification is a fancy way of saying that variable assignment must be consistent.
+
+
+== Multiple patterns
+
+In the previous example we received a `*fail*` when there was no match.
+If we instead want `false` instead we can add a catch all pattern.
+
+[source,clojure]
+----
+(def a
+  (m/rewrite
+    [?x ?x]
+    true
+    _
+    false))
+
+(a [1 1])
+;=> true
+(a [1 2])
+;=> false
+----
+
+Rewrite will accept multiple pattern pairs.
+
+`_` is a special match anything variable.
+
+
+== Sequences
+
+[source,clojure]
+----
+(def i
+  (m/rewrite
+    [!x ...]
+    [!x ...]))
+----
+
+`?x` means one, `!x` means many.
+`!x` is a memory variable.
+A memory variable means that there can be more than one value associated with it.
+An array of values will be collected instead of a single value.
+
+`...` means match 0 or more occurrences.
+We can think of `...` as being similar to the regex `*` operator.
+If we want 3 or more occurences we can use `..3` to specify a minimum number of matches.
+
+This is our first peek into an important principle in Meander; symmetry.
+When we use a pattern as both the match pattern and the substitution pattern our rewrite function will do nothing.
+The input is returned unchanged.
+This seemingly mundane properties has profound implications.
+We use exactly the same pattern language for matching and for substitution.
+There is a single unified way to sketch the shape of consumption and construction.
+
+
+== Multi-part sequences
+
+[source,clojure]
+----
+(def j
+  (m/rewrite
+    [!a !b ...]
+    [!a ... !b ...]))
+
+(j [1 2 3 4 5 6])
+;=> [1 3 5 2 4 6]
+----
+
+Notice that we do not need to think about the types of collections we will be producing.
+It is obvious that the collection is preserved.
+
+
+== Splitting sequences
+
+[source,clojure]
+----
+(def j
+  (m/rewrite
+    [?a . !b ...]
+    [!b ... ?a]))
+
+(j [1 2 3 4 5 6])
+;=> [2 3 4 5 6 1]
+----
+
+The `.` operator serves as a break to indicate how parts of the sequence should be grouped.
+
+
+== Rearranging an expression
+
+
+== Map nil punning
+
+[source,clojure]
+----
+(def h
+  (m/rewrite
+    {:foo ?v & ?rest-map}
+    [?v ?rest-map]))
```

Comment:
Am I miss understanding how this works?
```
(def h 
  (meander/rewrite 
    {:foo ?v & ?rest-map}
    [?v ?rest-map]))


Syntax error macroexpanding meander.epsilon/rewrite at (core.clj:2:3).
() - failed: Insufficient input at: [:clauses :substitution]
```

Reply:
This tutorial is very out of date. We should close this. We do have more documentation now on these sorts of things.